edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs

```diff
@@ -238,9 +249,57 @@ async void OnConnectionStatusChanged(ConnectionStatus status, ConnectionStatusCh
 
                 if (this.pullOnReconnect && this.initTask.IsCompleted && status == ConnectionStatus.Connected)
                 {
+                    var delayedTwinPull = true;
                     using (await this.twinLock.LockAsync())
                     {
-                        await this.RefreshTwinAsync();
+                        var now = DateTime.Now;
+                        if (now - this.lastTwinPullOnConnect > this.TwinPullOnConnectThrottleTime && !this.isDelayedTwinPullInProgress.Get())
+                        {
+                            this.lastTwinPullOnConnect = now;
+                            await this.RefreshTwinAsync();
+                            delayedTwinPull = false;
+                        }
+                    }
+
+                    if (delayedTwinPull)
+                    {
+                        if (this.isDelayedTwinPullInProgress.GetAndSet(true))
+                        {
+                            this.pullRequestCounter++;
+                        }
+                        else
+                        {
+                            _ = Task.Run(
+                                async () =>
+                                {
+                                    await Task.Delay(this.TwinPullOnConnectThrottleTime);
+
+                                    var requestCounter = default(int);
+                                    using (await this.twinLock.LockAsync())
+                                    {
+                                        this.lastTwinPullOnConnect = DateTime.Now;
+
+                                        try
+                                        {
+                                            await this.RefreshTwinAsync();
+                                        }
+                                        catch
```

Comment:
Might have missed it in the below tests, but if not, would you add a case around us having an exception on RefreshTwinAsync and we just reach the point where print out the warning?